### PR TITLE
[AIRFLOW-XXXX] Document new session handling guidelines

### PR DESCRIPTION
**Pull request by ashb**
---
Issue link: `Document only change, no JIRA issue`

Make sure to mark the boxes below before creating PR: [x]

- [x] Description above provides context of the change
- [x] Commit message/PR title starts with `[AIRFLOW-NNNN]`. AIRFLOW-NNNN = JIRA ID<sup>*</sup>
- [x] Unit tests coverage for changes (not needed for documentation changes)
- [x] Commits follow "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)"
- [x] Relevant documentation is updated including usage instructions.
- [x] I will engage committers as explained in [Contribution Workflow Example](https://github.com/apache/airflow/blob/master/CONTRIBUTING.rst#contribution-workflow-example).

<sup>*</sup> For document-only changes commit message can start with `[AIRFLOW-XXXX]`.

---
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/master/UPDATING.md).
Read the [Pull Request Guidelines](https://github.com/apache/airflow/blob/master/CONTRIBUTING.rst#pull-request-guidelines) for more information.

### CODE REVIEW

File: CONTRIBUTING.rst [hunk 1]
```diff
@@ -142,6 +142,11 @@ these guidelines:
     It will help you make sure you do not break the build with your PR and
     that you help increase coverage.
 
+-   Follow our project's `Coding style and best practices`_.
+
+    These are things that aren't currently enforced programtically (either because they are two hard or just
```
Comment (ashb) [inline CONTRIBUTING.rst]: Bah.

Comment (ashb) [inline CONTRIBUTING.rst]:
```suggestion
    These are things that aren't currently enforced programtically (either because they are too hard or just
```

**Review by potiuk — APPROVED** [review]
LGTM!

File: CONTRIBUTING.rst [hunk 2]
```diff
@@ -321,6 +326,47 @@ Your code must pass all the static code checks in Travis CI in order to be eligi
...
+-------------------------
+
+**Explicit is better than implicit.** If a function accepts a ``session`` parameter it should not commit the
+transaction itself. Session management is up to the caller.
+
+To make this easier there is the ``create_session`` helper:
+
+.. code-block:: python
+
+    from airflow.utils.db import create_session
```
Comment (kaxil) [inline CONTRIBUTING.rst]:
```suggestion
    from airflow.utils.session import create_session
```

File: CONTRIBUTING.rst [hunk 3]
```diff
@@ -321,6 +326,47 @@ Your code must pass all the static code checks in Travis CI in order to be eligi
...
+
+If this function is designed to be called by "end-users" (i.e. DAG authors) then using the ``@provide_session`` wrapper is okay:
+
+.. code-block:: python
+
+    from airflow.utils.db import provide_session
+
+    ...
+
+    @session
```
Comment (kaxil) [inline CONTRIBUTING.rst]:
```suggestion
    @provide_session
```

File: CONTRIBUTING.rst [hunk 4]
```diff
@@ -321,6 +326,47 @@ Your code must pass all the static code checks in Travis CI in order to be eligi
...
+      # You MUST not commit the session here.
+
+    with create_session() as session:
+        my_call(*args, session=session)
+
+If this function is designed to be called by "end-users" (i.e. DAG authors) then using the ``@provide_session`` wrapper is okay:
+
+.. code-block:: python
+
+    from airflow.utils.db import provide_session
```
Comment (kaxil) [inline CONTRIBUTING.rst]:
```suggestion
    from airflow.utils.session import provide_session
```

File: CONTRIBUTING.rst [hunk 5]
```diff
@@ -321,6 +326,47 @@ Your code must pass all the static code checks in Travis CI in order to be eligi
 The easiest way to make sure your code is good before pushing is to use pre-commit checks locally
 as described in the static code checks documentation.
 
+.. _coding_style:
+
+Coding style and best practices
+===============================
+
+Most of your coding style rules are enforced programmatically by flake8 and pylint (which are run automatically
```
Comment (kaxil) [inline CONTRIBUTING.rst]:
```suggestion
Most of our coding style rules are enforced programmatically by flake8 and pylint (which are run automatically
```

**Comment by ashb** [comment]
Lol thanks Kaxil. Rushed PRs are never worth it :D

**Review by kaxil — APPROVED** [review]
😄 LGTM

**Review by yuqian90 — APPROVED** [review]

**Comment by yuqian90** [comment]
Lgtm thanks!

**Comment by nuclearpinguin** [comment]
@ash should we link those best practices in `firstPRWelcomeComment` in boring cyborg?

**Comment by ashb** [comment]
Good idea Tomek, will open a PR there too

**Review by Fokko — APPROVED** [review]
LGTM

**Review by turbaszek — APPROVED** [review]

**Pull request merged**